Add page on OpenSearch keystore in security configuration (#6304)
opensearch-project/documentation-website: _security/configuration/opensearch-keystore.md

---
layout: default
title: OpenSearch keystore
parent: Configuration
nav_order: 50
---

> # OpenSearch keystore

Reviewer: Should we change this and the title to "OpenSearch keystore script"?

Author: IMO it may make sense to have a dropdown for documentation of all OpenSearch scripts [scripts](https://github.com/opensearch-project/OpenSearch/tree/main/distribution/src/bin) and then include pages for `OpenSearch keystore`, `OpenSearch plugin`, etc. I included this in security configuration because its related to storing secrets securely (i.e. encrypted in the keystore instead of in plain text in files)

Note, this isn't a feature of the security plugin, this is part of the core of OpenSearch.

Reviewer 2: Lets keep this here for now. I think people are more likely to look for this in the security documentation versus a "script" section, but definitely something to keep in mind for later.

`opensearch-keystore` is a utility script for managing the OpenSearch keystore. This keystore provides a secure method to store sensitive information, such as passwords and keys, used in an OpenSearch cluster. The script allows you to securely create, list, add, and remove settings. It is included inside the OpenSearch distribution. 

## Usage

> In order to use the OpenSearch keystore script, you must have access to the file system in which the script resides. 

Author: This is the same file system that the opensearch installation is in. Is there a better way to phrase this? A user who has access to execute OpenSearch scripts can start OpenSearch, upgrade OpenSearch, install plugins and set settings. They would have access to all scripts [here](https://github.com/opensearch-project/OpenSearch/tree/main/distribution/src/bin).

> In order to use the OpenSearch keystore script, you must have access to the file system in which the script resides. 

Reviewer: I would probably just keep it simple and name the installation: "In order to use the `opensearch-keystore` script, you must have access to [Name](link)."

> In order to use the OpenSearch keystore script, you must have access to the file system in which the script resides. 

Author: wdyt? 

In order to use the `opensearch-keystore` script, you must have access to the file system with the OpenSearch installation and the ability to execute OpenSearch scripts.

> In order to use the OpenSearch keystore script, you must have access to the file system in which the script resides. 

Reviewer: Yes, but `containing` instead of `with`.

> To use OpenSearch keystore, open a terminal and use the following command syntax: 

Author: wdyt?

To use `opensearch-keystore`, open a terminal and use the following command syntax:

> To use OpenSearch keystore, open a terminal and use the following command syntax: 

Reviewer: Works for me 👍

```
opensearch-keystore [command] [options]
```
{% include copy.html %}

## Commands

OpenSearch keystore supports the following the commands: 

- `create`: Initializes a new keystore. If a keystore already exists, this command will overwrite the existing keystore.

> - `list`: Lists all settings in the keystore.

Reviewer: ```suggestion
- `list`: Lists all keystore settings.
```

Author: IMO I think `Lists all settings in the keystore.` is the way to go here. i.e. You list all files in a directory and not list all directory files.

- `add <setting-name>`: Adds a new setting to the current keystore. When a new setting is added, OpenSearch keystore prompts for the value of that setting. After adding the setting and value, both are securely stored in the keystore.
- `add-file <file-name>`: Adds a new file to the keystore.
- `remove <setting-name>`: Removes an existing setting from the keystore.
- `upgrade <setting-name>`: Upgrades an existing setting in the keystore.
- `passwd`: Sets a password for the keystore.

> - `has-passwd`: Prints whether the keystore is password protected or not.

Reviewer: ```suggestion
- `has-passwd`: Prints whether the keystore is password protected.
```

> - `has-passwd`: Prints whether the keystore is password protected or not.

Reviewer: Is "Prints" the right word, or would "Indicates" work?

- `help`: Displays help information about all `opensearch-keystore` commands.

## Options

You can append each command with the following options:

> - `-h, --help`: Displays help information about the keystore script and its options.

Reviewer: ```suggestion
- `-h, --help`: Displays help information about the script and its options.
```

> - `-s, --silent`: Provides minimal output when OpenSearch keystore responds to a command.

Reviewer: ```suggestion
- `-s, --silent`: Provides minimal output when the script responds to a command.
```

- `-v, --verbose`: Provides a verbose output for debugging purposes.

## Examples

> The following examples provide the basic syntax for common OpenSearch keystore commands:

Reviewer: ```suggestion
The following examples provide the basic syntax for common `opensearch-keystore` commands:
```

> ### Creating a new keystore:

Reviewer: ```suggestion
### Creating a new keystore
```

> ### Creating a new keystore:

Reviewer: No terminal punctuation in headings 😄.

**Command**

```bash
./bin/opensearch-keystore create
```
{% include copy.html %}

> If a keystore already exists, the script will ask if you like to overwrite the existing keystore.

Reviewer: ```suggestion
If a keystore already exists, the script will ask whether you would like to overwrite the existing keystore.
```

   
**Response**

> The script responds with confirmation that the keystore was created:

Reviewer: ```suggestion
The script responds with a confirmation that the keystore was created:
```

   
```bash

> Created opensearch keystore in /Users/cwperx/Projects/opensearch/OpenSearch/build/distribution/local/opensearch-2.11.2-SNAPSHOT/config/opensearch.keystore

Author: ```suggestion
Created opensearch keystore in $OPENSEARCH_HOME/config/opensearch.keystore
```

```

> ### Listing settings in the keystore:

Reviewer: ```suggestion
### Listing keystore settings
```

Author: Same here. I will remove the colon.

**Command**
   
```bash
./bin/opensearch-keystore list
```
{% include copy.html %}

**Response**

> The script responds with a list of settings in the keystore:

Reviewer: ```suggestion
The script responds with a list of keystore settings:
```

Author: Leaving as is.

```bash
keystore.seed
plugins.security.ssl.http.pemkey_password_secure
```

> ### Adding a new setting:

Reviewer: ```suggestion
### Adding a new setting
```

```bash
./bin/opensearch-keystore add plugins.security.ssl.http.pemkey_password_secure
```
{% include copy.html %}

**Response**

After this command, you will be prompted to enter the secret key securely.

> ### Removing a setting:

Reviewer: ```suggestion
### Removing a setting
```

**Command**

```bash
./bin/opensearch-keystore remove plugins.security.ssl.http.pemkey_password_secure
```
{% include copy-curl.html %}

**Response**

> No response exists for this command. To confim the setting was deleted, use `opensearch-keystore list`.

Reviewer 2: ```suggestion
No response exists for this command. To confirm the setting was deleted, use `opensearch-keystore list`.
```

> No response exists for this command. To confirm the setting was deleted, use `opensearch-keystore list`.

Reviewer: ```suggestion
No response exists for this command. To confirm that the setting was deleted, use `opensearch-keystore list`.
```

## Referring to keystore entries

> After a setting has been added to the keystore, you can refer back to that setting inside your OpenSearch configuration. To refer back to the setting, place the keystore setting name as a placeholder in the `opensearch.yml` configuration file, as shown in the following example:

Reviewer: ```suggestion
After a setting has been added to a keystore, you can refer back to that setting in your OpenSearch configuration. To refer back to the setting, place the keystore setting name as a placeholder in the `opensearch.yml` configuration file, as shown in the following example:
```

> After a setting has been added to the keystore, you can refer back to that setting inside your OpenSearch configuration. To refer back to the setting, place the keystore setting name as a placeholder in the `opensearch.yml` configuration file, as shown in the following example:

Reviewer: Can we use a word other than "place" in order to avoid place/placeholder? "store"?

Author: `add`, `enter` or `put`?

Reviewer: `add`

---
layout: default

> title: OpenSearch keystore

Reviewer 2: Global: OpenSearch keystore or OpenSearch Keystore? I'll use the first for now but let me know.

Author: Switched this to capitalize Keystore for consistency.

Reviewer: It should only be capitalized as a proper noun, in other words, if this is a formally named feature that is specific to OpenSearch.

Author: @natebower wdyt in this case? This is one of the scripts that is shipped with [opensearch](https://github.com/opensearch-project/OpenSearch/tree/main/distribution/src/bin). Here's another [page](https://opensearch.org/docs/latest/install-and-configure/plugins/) on the documentation-website that details the `opensearch-plugin` script.

Reviewer: Based on the content at the links, I would probably use "OpenSearch keystore" as the title and heading and when referencing _the_ OpenSearch keystore and `opensearch-keystore` (as you've done above) when referencing the name of the script.

Author: 👍 I updated all instances from OpenSearch Keystore to OpenSearch keystore in this doc.

parent: Configuration
nav_order: 50
---

# OpenSearch Keystore

> `opensearch-keystore` is a utility script for managing the OpenSearch keystore. This keystore provides a secure method to store sensitive information, such as passwords and keys, used in the OpenSearch cluster. The script allows users to create, list, add, and remove settings securely.

Reviewer 2: ```suggestion
`opensearch-keystore` is a utility script for managing the OpenSearch Keystore. This keystore provides a secure method to store sensitive information, such as passwords and keys, used in an OpenSearch cluster. The script allows you to securely create, list, add, and remove settings.
```

> ## Usage

Reviewer 2: ```suggestion
## Usage

To use OpenSearch Keystore, open a terminal and use the following command syntax: 

```

```
opensearch-keystore [command] [options]
```

> ### Commands

Reviewer 2: ```suggestion
## Commands
```

> 

Reviewer 2: ```suggestion
OpenSearch Keystore supports the following the commands: 

```

> - `create`: Initializes a new keystore. If a keystore already exists, this command will overwrite it.

Reviewer 2: ```suggestion
- `create`: Initializes a new keystore. If a keystore already exists, this command will overwrite the existing keystore.
```

> - `list`: Lists all settings in the keystore.

Reviewer 2: Is there a place I can find a list of settings I can to the keystore?

Author: The only place you can see a full list is in the source code [here](https://github.com/opensearch-project/OpenSearch/blob/main/distribution/tools/keystore-cli/src/main/java/org/opensearch/common/settings/KeyStoreCli.java#L43-L53). This [article](https://opster.com/guides/opensearch/opensearch-security/opensearch-keystore/) from Opster is also a good reference for usage of the opensearch-keystore tool.

> - `add <setting-name>`: Adds a new setting to the keystore. Prompts for the value to be securely stored.

Reviewer 2: ```suggestion
- `add <setting-name>`: Adds a new setting to the current keystore. When a new setting is added, OpenSearch Keystore prompts for the value of that setting. After adding the setting and value, both are securely stored in the keystore.
```

> - `add-file <file-name>`: Adds a new file to the keystore.

Reviewer 2: What type of files would you add to the keystore?

Author: Private keys for TLS would be an example I could think of where the contents of the file should be kept secret.

Reviewer: Ok, I'm fine with that on lines 68 and 79 as well, then.

> - `remove <setting-name>`: Removes a setting from the keystore.

Reviewer 2: ```suggestion
- `remove <setting-name>`: Removes an existing setting from the keystore.
```

> - `upgrade <setting-name>`: Upgrades an existing setting within the keystore.

Reviewer 2: ```suggestion
- `upgrade <setting-name>`: Upgrades an existing setting in the keystore.
```

> - `passwd`: Sets a password for the keystore.

Reviewer 2: IS there a specific permission required to access the keystore or just a password?

Author: You need permission to execute the `opensearch-keystore` script as well which requires access to the filesystem for where the script resides.

Reviewer 2: How does a user download the script?

Author: Its included in the distribution. There's no need to download.

> - `has-passwd`: Print whether keystore is password protected or not.

Reviewer 2: ```suggestion
- `has-passwd`: Prints whether the keystore is password protected or not.
```

> - `help`: Displays help information about the `opensearch-keystore` commands.

Reviewer 2: ```suggestion
- `help`: Displays help information about all `opensearch-keystore` commands.
```

> ### Options

Reviewer 2: ```suggestion
## Options
```

> 

Reviewer 2: ```suggestion

You can append each command with the following options:

```

> - `-h, --help`: Displays help information about the script and its options.

Reviewer 2: ```suggestion
- `-h, --help`: Displays help information about the keystore script and its options.
```

> - `-s, --silent`: Provides minimal output.

Reviewer 2: ```suggestion
- `-s, --silent`: Provides minimal output when OpenSearch Keystore responds to a command.
```

- `-v, --verbose`: Provides verbose output for debugging purposes.

### Examples

> 1. Creating a new keystore:

Author: 1. Create

```
> ./bin/opensearch-keystore create
An opensearch keystore already exists. Overwrite? [y/N]y
Created opensearch keystore in /Users/cwperx/Projects/opensearch/OpenSearch/build/distribution/local/opensearch-2.11.2-SNAPSHOT/config/opensearch.keystore
```
2. List

```
> ./bin/opensearch-keystore list
keystore.seed
plugins.security.ssl.http.pemkey_password_secure
```

3. Add

```
> ./bin/opensearch-keystore add plugins.security.ssl.http.pemkey_password_secure
Enter value for plugins.security.ssl.http.pemkey_password_secure:
```

4. Remove

```
> ./bin/opensearch-keystore remove plugins.security.ssl.http.pemkey_password_secure

No output, run list again to see that the setting has been removed
```
